src/util_regression.mjs

```diff
@@ -150,7 +152,9 @@ util_regression.ridge = function(y, X, k){
             if (m_Coefficients.length*n !== m_Coefficients.length){
                 console.log('Array length must be a multiple of m')
             }
-            solution = (ss.length === ss[0].length ? (numeric.LUsolve(numeric.LU(ss,true),bb)) : (webgazer.mat.QRDecomposition(ss,bb)));
+            solution = ss.length === ss[0].length ?
+              mathjs.squeeze(mathjs.lusolve(ss, bb)) : // lusolve returns a column vector, so squeeze it
+              webgazer.mat.QRDecomposition(ss,bb);
```

Comment:
math.js has a qr decomposition too, right?

Reply:
Right, but the [return value](https://mathjs.org/docs/reference/functions/qr.html) is different from `mat.QRDecomposition()`

Reply:
The `qrsolve()` function can be implemented as:
```js
function qrsolve(A, b) {
  const { Q, R } = math.qr(A);
  const dim1 = A[0].length;
  
  return math.multiply(
    math.inv(R.slice(0, dim1)),
    math.transpose(Q).slice(0, dim1),
    b,
  );
}
```
However, I don't think it will be faster than the current one.

Reply:
Those implements in `mat.mjs` seem to be much faster than `mathjs` in my tests:

![image](https://user-images.githubusercontent.com/6022672/122142198-ebfd0080-ce89-11eb-9d0f-595377dc9694.png)

[Test Code](https://gist.github.com/idiotWu/80de6019ee80a3e6c4d6ddb19f2eb7b5)

Reply:
WOW. Just wow. Add a comment so we don't accidentally use them in the future. We should verify the matmul and other functions in math.js are actually faster.

Reply:
`mat.mul()` is also faster than `mathjs.multiply()` after 1M runs:

![image](https://user-images.githubusercontent.com/6022672/122143290-29628d80-ce8c-11eb-8012-9fcb246d1d2b.png)

Maybe mathjs is wasting too much time constructing matrix instances 😞

Reply:
Rather disappointing. Are there any other good candidates that may work better than math.js. Surely something has to be faster than numeric, maybe something that even leveraged WASM?

Reply:
In my tests, mat.* >>> numeric ~ [ml-matrix](https://github.com/mljs/matrix) >> [math.js](https://github.com/josdejong/mathjs). Maybe the best solution is to implement all the required matrix operations ourselves.

Reply:
> `mat.mul()` is also faster than `mathjs.multiply()` after 1M runs:
> 
> ![image](https://user-images.githubusercontent.com/6022672/122143290-29628d80-ce8c-11eb-8012-9fcb246d1d2b.png)
> 
> Maybe mathjs is wasting too much time constructing matrix instances 😞

Looked into the code and apparently the matrix instances are just a wrapper around associative arrays so that shouldn't slow it down. I did see that it did have serious speed issues though. It's designed to be flexible and extensible, not fast.